File: obligation-cordapp/java-source/src/test/java/net/corda/examples/obligation/flows/SettleObligationTests.java
package net.corda.examples.obligation.flows;

import net.corda.core.contracts.Amount;
import net.corda.core.flows.FlowException;
import net.corda.core.identity.Party;
import net.corda.core.transactions.SignedTransaction;
import net.corda.examples.obligation.Obligation;
import net.corda.finance.contracts.asset.Cash;
import net.corda.testing.node.StartedMockNode;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import java.util.Currency;
import java.util.List;

import static net.corda.core.contracts.Structures.withoutIssuer;
import static net.corda.finance.Currencies.POUNDS;
import static net.corda.testing.internal.InternalTestUtilsKt.chooseIdentity;
import static org.hamcrest.CoreMatchers.instanceOf;
import static org.junit.Assert.assertEquals;

public class SettleObligationTests extends ObligationTests {

    // Helper for extracting the cash output owned by a the node.
    private Cash.State getCashOutputByOwner(
            List<Cash.State> cashStates,
            StartedMockNode node) {
        Cash.State ownersCashState = null;
        for (Cash.State cashState : cashStates) {
            Party cashOwner = node.getServices().getIdentityService().requireWellKnownPartyFromAnonymous(cashState.getOwner());
            if (cashOwner == chooseIdentity(node.getInfo())) {
                ownersCashState = cashState;
                break;
            }
        }
        return ownersCashState;
    }

    @Rule
    public final ExpectedException exception = ExpectedException.none();

    @Test
    public void settleFlowCanOnlyBeStartedByBorrower() throws Exception {
        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Attempt settlement.
        exception.expectCause(instanceOf(FlowException.class));
        settleObligation(issuedObligation.getLinearId(), b, POUNDS(1000), true);
    }

    @Test
    public void settleFlowFailsWhenBorrowerHasNoCash() throws Exception {
        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Attempt settlement.
        exception.expectCause(instanceOf(FlowException.class));
        settleObligation(issuedObligation.getLinearId(), a, POUNDS(1000), true);
    }

    @Test
    public void settleFlowFailsWhenBorrowerPledgesTooMuchCashToSettle() throws Exception {
        // Self issue cash.
        selfIssueCash(a, POUNDS(1500));
        network.waitQuiescent();

        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Attempt settlement.
        exception.expectCause(instanceOf(FlowException.class));
        settleObligation(issuedObligation.getLinearId(), a, POUNDS(1500), true);
    }

    @Test
    public void fullySettleNonAnonymousObligation() throws Exception {
        // Self issue cash.
        selfIssueCash(a, POUNDS(1500));
        network.waitQuiescent();

        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Attempt settlement.
        SignedTransaction settleTransaction = settleObligation(issuedObligation.getLinearId(), a, POUNDS(1000), true);
        network.waitQuiescent();
        assert(settleTransaction.getTx().outputsOfType(Obligation.class).isEmpty());

        // Check both parties have the transaction.
        SignedTransaction aTx = a.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        SignedTransaction bTx = b.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        assertEquals(aTx, bTx);
    }

    @Test
    public void fullySettleAnonymousObligation() throws Exception {
        // Self issue cash.
        selfIssueCash(a, POUNDS(1500));

        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), true);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Attempt settlement.
        SignedTransaction settleTransaction = settleObligation(issuedObligation.getLinearId(), a, POUNDS(1000), true);
        network.waitQuiescent();
        assert(settleTransaction.getTx().outputsOfType(Obligation.class).isEmpty());

        // Check both parties have the transaction.
        SignedTransaction aTx = a.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        SignedTransaction bTx = b.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        assertEquals(aTx, bTx);
    }

    @Test
    public void partiallySettleNonAnonymousObligationWithNonAnonymousCashPayment() throws Exception {
        // Self issue cash.
        selfIssueCash(a, POUNDS(1500));
        network.waitQuiescent();

        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Attempt settlement.
        Amount<Currency> amountToSettle = POUNDS(500);
        SignedTransaction settleTransaction = settleObligation(issuedObligation.getLinearId(), a, amountToSettle, false);
        network.waitQuiescent();
        assertEquals(1, settleTransaction.getTx().outputsOfType(Obligation.class).size());

        // Check both parties have the transaction.
        SignedTransaction aTx = a.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        SignedTransaction bTx = b.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        assertEquals(aTx, bTx);

        // Check the obligation paid amount is correctly updated.
        Obligation partiallySettledObligation = settleTransaction.getTx().outputsOfType(Obligation.class).get(0);
        assertEquals(amountToSettle, partiallySettledObligation.getPaid());

        // Check cash has gone to the correct parties.
        List<Cash.State> outputCash = settleTransaction.getTx().outputsOfType(Cash.State.class);
        assertEquals(2, outputCash.size());       // Cash to b and change to a.

        // Change addresses are always anonymous, I think.
        Cash.State change = getCashOutputByOwner(outputCash, a);
        assertEquals(POUNDS(1000), withoutIssuer(change.getAmount()));

        Cash.State payment = getCashOutputByOwner(outputCash, b);
        assertEquals(POUNDS(500), withoutIssuer(payment.getAmount()));
    }

    @Test
    public void partiallySettleNonAnonymousObligationWithAnonymousCashPayment() throws Exception {
        // Self issue cash.
        selfIssueCash(a, POUNDS(1500));
        network.waitQuiescent();

        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), false);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Attempt settlement.
        Amount<Currency> amountToSettle = POUNDS(500);
        SignedTransaction settleTransaction = settleObligation(issuedObligation.getLinearId(), a, amountToSettle, true);
        network.waitQuiescent();
        assertEquals(1, settleTransaction.getTx().outputsOfType(Obligation.class).size());

        // Check both parties have the transaction.
        SignedTransaction aTx = a.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        SignedTransaction bTx = b.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        assertEquals(aTx, bTx);

        // Check the obligation paid amount is correctly updated.
        Obligation partiallySettledObligation = settleTransaction.getTx().outputsOfType(Obligation.class).get(0);
        assertEquals(amountToSettle, partiallySettledObligation.getPaid());

        // Check cash has gone to the correct parties.
        List<Cash.State> outputCash = settleTransaction.getTx().outputsOfType(Cash.State.class);
        assertEquals(2, outputCash.size());       // Cash to b and change to a.

        Cash.State change = getCashOutputByOwner(outputCash, a);
        assertEquals(POUNDS(1000), withoutIssuer(change.getAmount()));

        Cash.State payment = getCashOutputByOwner(outputCash, b);
        assertEquals(POUNDS(500), withoutIssuer(payment.getAmount()));
    }

    @Test
    public void partiallySettleAnonymousObligationWithAnonymousCashPayment() throws Exception {
        // Self issue cash.
        selfIssueCash(a, POUNDS(1500));
        network.waitQuiescent();

        // Issue obligation.
        SignedTransaction issuanceTransaction = issueObligation(a, b, POUNDS(1000), true);
        network.waitQuiescent();
        Obligation issuedObligation = (Obligation) issuanceTransaction.getTx().getOutputStates().get(0);

        // Attempt settlement.
        Amount<Currency> amountToSettle = POUNDS(500);
        SignedTransaction settleTransaction = settleObligation(issuedObligation.getLinearId(), a, amountToSettle, true);
        network.waitQuiescent();
        assertEquals(1, settleTransaction.getTx().outputsOfType(Obligation.class).size());

        // Check both parties have the transaction.
        SignedTransaction aTx = a.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        SignedTransaction bTx = b.getServices().getValidatedTransactions().getTransaction(settleTransaction.getId());
        assertEquals(aTx, bTx);

        // Check the obligation paid amount is correctly updated.
        Obligation partiallySettledObligation = settleTransaction.getTx().outputsOfType(Obligation.class).get(0);
        assertEquals(amountToSettle, partiallySettledObligation.getPaid());

        // Check cash has gone to the correct parties.
        List<Cash.State> outputCash = settleTransaction.getTx().outputsOfType(Cash.State.class);
        assertEquals(2, outputCash.size());       // Cash to b and change to a.

        Cash.State change = getCashOutputByOwner(outputCash, a);
        assertEquals(POUNDS(1000), withoutIssuer(change.getAmount()));

        Cash.State payment = getCashOutputByOwner(outputCash, b);
        assertEquals(POUNDS(500), withoutIssuer(payment.getAmount()));
    }
}

File: obligation-cordapp/kotlin-source/src/test/kotlin/net/corda/examples/obligation/flows/SettleObligationTests.kt
package net.corda.examples.obligation.flows

import net.corda.core.contracts.withoutIssuer
import net.corda.core.flows.FlowException
import net.corda.examples.obligation.Obligation
import net.corda.finance.POUNDS
import net.corda.finance.contracts.asset.Cash
import net.corda.testing.internal.chooseIdentity
import net.corda.testing.node.StartedMockNode
import kotlin.test.*

class SettleObligationTests : ObligationTests() {

    // Helper for extracting the cash output owned by a the node.
    private fun getCashOutputByOwner(
            cashStates: List<Cash.State>,
            node: StartedMockNode): Cash.State {
        return cashStates.single { cashState ->
            val cashOwner = node.services.identityService.requireWellKnownPartyFromAnonymous(cashState.owner)
            cashOwner == node.info.chooseIdentity()
        }
    }

    @org.junit.Test
    fun `Settle flow can only be started by borrower`() {
        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS, anonymous = false)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Attempt settlement.
        assertFailsWith<FlowException> {
            settleObligation(issuedObligation.linearId, b, 1000.POUNDS)
        }
    }

    @org.junit.Test
    fun `Settle flow fails when borrower has no cash`() {
        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS, anonymous = false)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Attempt settlement.
        assertFailsWith<FlowException> {
            settleObligation(issuedObligation.linearId, a, 1000.POUNDS)
        }
    }

    @org.junit.Test
    fun `Settle flow fails when borrower pledges too much cash to settle`() {
        // Self issue cash.
        selfIssueCash(a, 1500.POUNDS)
        network.waitQuiescent()

        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS, anonymous = false)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Attempt settlement.
        assertFailsWith<FlowException> {
            settleObligation(issuedObligation.linearId, a, 1500.POUNDS)
        }
    }

    @org.junit.Test
    fun `Fully settle non-anonymous obligation`() {
        // Self issue cash.
        selfIssueCash(a, 1500.POUNDS)
        network.waitQuiescent()

        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS, anonymous = false)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Attempt settlement.
        val settleTransaction = settleObligation(issuedObligation.linearId, a, 1000.POUNDS)
        network.waitQuiescent()
        assert(settleTransaction.tx.outputsOfType<Obligation>().isEmpty())

        // Check both parties have the transaction.
        val aTx = a.services.validatedTransactions.getTransaction(settleTransaction.id)
        val bTx = b.services.validatedTransactions.getTransaction(settleTransaction.id)
        assertEquals(aTx, bTx)
    }

    @org.junit.Test
    fun `Fully settle anonymous obligation`() {
        // Self issue cash.
        selfIssueCash(a, 1500.POUNDS)

        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Attempt settlement.
        val settleTransaction = settleObligation(issuedObligation.linearId, a, 1000.POUNDS)
        network.waitQuiescent()
        assert(settleTransaction.tx.outputsOfType<Obligation>().isEmpty())

        // Check both parties have the transaction.
        val aTx = a.services.validatedTransactions.getTransaction(settleTransaction.id)
        val bTx = b.services.validatedTransactions.getTransaction(settleTransaction.id)
        assertEquals(aTx, bTx)
    }

    @org.junit.Test
    fun `Partially settle non-anonymous obligation with non-anonymous cash payment`() {
        // Self issue cash.
        selfIssueCash(a, 1500.POUNDS)
        network.waitQuiescent()

        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS, anonymous = false)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Attempt settlement.
        val amountToSettle = 500.POUNDS
        val settleTransaction = settleObligation(issuedObligation.linearId, a, amountToSettle, anonymous = false)
        network.waitQuiescent()
        assertEquals(1, settleTransaction.tx.outputsOfType<Obligation>().size)

        // Check both parties have the transaction.
        val aTx = a.services.validatedTransactions.getTransaction(settleTransaction.id)
        val bTx = b.services.validatedTransactions.getTransaction(settleTransaction.id)
        assertEquals(aTx, bTx)

        // Check the obligation paid amount is correctly updated.
        val partiallySettledObligation = settleTransaction.tx.outputsOfType<Obligation>().single()
        assertEquals(amountToSettle, partiallySettledObligation.paid)

        // Check cash has gone to the correct parties.
        val outputCash = settleTransaction.tx.outputsOfType<Cash.State>()
        assertEquals(2, outputCash.size)       // Cash to b and change to a.

        // Change addresses are always anonymous, I think.
        val change = getCashOutputByOwner(outputCash, a)
        assertEquals(1000.POUNDS, change.amount.withoutIssuer())

        val payment = getCashOutputByOwner(outputCash, b)
        assertEquals(500.POUNDS, payment.amount.withoutIssuer())
    }

    @org.junit.Test
    fun `Partially settle non-anonymous obligation with anonymous cash payment`() {
        // Self issue cash.
        selfIssueCash(a, 1500.POUNDS)
        network.waitQuiescent()

        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS, anonymous = false)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Attempt settlement.
        val amountToSettle = 500.POUNDS
        val settleTransaction = settleObligation(issuedObligation.linearId, a, amountToSettle)
        network.waitQuiescent()
        assert(settleTransaction.tx.outputsOfType<Obligation>().size == 1)

        // Check both parties have the transaction.
        val aTx = a.services.validatedTransactions.getTransaction(settleTransaction.id)
        val bTx = b.services.validatedTransactions.getTransaction(settleTransaction.id)
        assertEquals(aTx, bTx)

        // Check the obligation paid amount is correctly updated.
        val partiallySettledObligation = settleTransaction.tx.outputsOfType<Obligation>().single()
        assertEquals(amountToSettle, partiallySettledObligation.paid)

        // Check cash has gone to the correct parties.
        val outputCash = settleTransaction.tx.outputsOfType<Cash.State>()
        assertEquals(2, outputCash.size)       // Cash to b and change to a.

        val change = getCashOutputByOwner(outputCash, a)
        assertEquals(1000.POUNDS, change.amount.withoutIssuer())

        val payment = getCashOutputByOwner(outputCash, b)
        assertEquals(500.POUNDS, payment.amount.withoutIssuer())
    }

    @org.junit.Test
    fun `Partially settle anonymous obligation with anonymous cash payment`() {
        // Self issue cash.
        selfIssueCash(a, 1500.POUNDS)
        network.waitQuiescent()

        // Issue obligation.
        val issuanceTransaction = issueObligation(a, b, 1000.POUNDS)
        network.waitQuiescent()
        val issuedObligation = issuanceTransaction.tx.outputStates.first() as Obligation

        // Attempt settlement.
        val amountToSettle = 500.POUNDS
        val settleTransaction = settleObligation(issuedObligation.linearId, a, amountToSettle)
        network.waitQuiescent()
        assertEquals(1, settleTransaction.tx.outputsOfType<Obligation>().size)

        // Check both parties have the transaction.
        val aTx = a.services.validatedTransactions.getTransaction(settleTransaction.id)
        val bTx = b.services.validatedTransactions.getTransaction(settleTransaction.id)
        assertEquals(aTx, bTx)

        // Check the obligation paid amount is correctly updated.
        val partiallySettledObligation = settleTransaction.tx.outputsOfType<Obligation>().single()
        assertEquals(amountToSettle, partiallySettledObligation.paid)

        // Check cash has gone to the correct parties.
        val outputCash = settleTransaction.tx.outputsOfType<Cash.State>()
        assertEquals(2, outputCash.size)       // Cash to b and change to a.

        val change = getCashOutputByOwner(outputCash, a)
        assertEquals(1000.POUNDS, change.amount.withoutIssuer())

        val payment = getCashOutputByOwner(outputCash, b)
        assertEquals(500.POUNDS, payment.amount.withoutIssuer())
    }

}
